### PR TITLE
Adimako/wave tools/cython

### PR DESCRIPTION
**Pull request by cekees**
This branch is for optimizing WaveTools and adding the ability to use arbitrary time series.

### CODE REVIEW

File: proteus/WaveTools.pyx [hunk 1]
```diff
@@ -60,7 +101,11 @@ def piersonMoskovitz(f,f0,alpha=8.1e-3,beta=0.74,g=9.8):
 def cos2s(theta,s):
     return cos(theta/2)**(2*s)
 
-def normInt(thetas,dir_fun,s,N):
+<<<<<<< HEAD:proteus/WaveTools.pyx
```
Comment (smattis) [inline proteus/WaveTools.pyx]: conflicts need resolved

File: proteus/WaveTools.pyx [hunk 2]
```diff
@@ -208,38 +241,24 @@ def __init__(self,period,waveHeight,mwl,depth,g,waveDir,wavelength=None,waveType
...
-            else: return HH
-    def Z(self,x,y,z):
-        return   -(self.vDir[0]*x + self.vDir[1]*y+ self.vDir[2]*z) - self.mwl
-
-    def u(self,x,y,z,t,ss = "x"):
-        """x-component of velocity
+                HH+=eta_mode(x,y,z,t,ii*self.kDir,ii*self.omega,self.phi0,Y)
+            return HH/self.k
 
+<<<<<<< HEAD:proteus/WaveTools.pyx
```
Comment (smattis) [inline proteus/WaveTools.pyx]: conflicts need resolved

File: proteus/WaveTools.pyx [hunk 3]
```diff
@@ -247,19 +266,17 @@ def u(self,x,y,z,t,ss = "x"):
         UH=0.
         UV=0.
         ii=0.
+        y=0
+=======
+    def u(self,x,y,z,t,comp):
+>>>>>>> adimako/WaveTools/cython:proteus/WaveTools.pyx
```
Comment (smattis) [inline proteus/WaveTools.pyx]: conflicts need resolved

File: proteus/WaveTools.pyx [hunk 4]
```diff
@@ -247,19 +266,17 @@ def u(self,x,y,z,t,ss = "x"):
...
-                "y":UH*self.waveDir[1] + UV*self.vDir[1],
-                "z":UH*self.waveDir[2] + UV*self.vDir[2],
-                }
-
+            return vel_mode(x,y,z,t,self.kDir,self.omega,self.phi0,self.amplitude,self.mwl,self.depth,self.g,comp)
         elif self.waveType is "Fenton":
+            U =0.
             for B in self.Bcoeff:
                 ii+=1
+<<<<<<< HEAD:proteus/WaveTools.pyx
```
Comment (smattis) [inline proteus/WaveTools.pyx]: conflicts need resolved

File: proteus/WaveTools.pyx [hunk 5]
```diff
@@ -270,10 +287,14 @@ def u(self,x,y,z,t,ss = "x"):
...
         else:
-            logEvent("Check Wave types. Available wave types are %s" % waveType,level=0)
+            logEvent("Check Wave types. Available wave types are %s" % (self.waveType,),level=0)
             exit(1)
         return Vcomp[ss]
 
+=======
+                U+=vel_mode(x,y,z,t,ii*self.kDir,ii*self.omega,self.phi0,ii*B,self.mwl,self.depth,self.g,comp)
+            return U+self.meanvelocity
+>>>>>>> adimako/WaveTools/cython:proteus/WaveTools.pyx
```
Comment (smattis) [inline proteus/WaveTools.pyx]: conflicts need resolved

File: proteus/WaveTools.pyx [hunk 6]
```diff
@@ -654,33 +678,62 @@ def reconstruct_direct(self,x,y,z,t,Nf,var="eta",ss = "x"):
...
+                return Eta
+            else:
+                return 0.
         if var=="U":
             UH=0.
             UV=0.
             for ii in range(Nf):
-                UH+=ai[ii]*omega[ii]*cosh(ki[ii]*(Z(x,y,z)+depth))*cos(x*kDir[ii,0]+y*kDir[ii,1]+z*kDir[ii,2] - omega[ii]*t + phi[ii])/sinh(ki[ii]*depth)
-                UV+=ai[ii]*omega[ii]*sinh(ki[ii]*(Z(x,y,z)+depth))*sin(x*kDir[ii,0]+y*kDir[ii,1]+z*kDir[ii,2] - omega[ii]*t + phi[ii])/sinh(ki[ii]*depth)
+<<<<<<< HEAD:proteus/WaveTools.pyx
```
Comment (smattis) [inline proteus/WaveTools.pyx]: conflicts need resolved

File: setup.py [hunk 7]
```diff
@@ -47,7 +47,14 @@
       packages = ['proteus', 'proteus.config', 'proteus.tests', 'proteus.mprans'],
       cmdclass = {'build_ext':build_ext},
       ext_package='proteus',
-      ext_modules=[Extension("ADR",['proteus/ADR.pyx'],
+      ext_modules=[Extension("WaveTools",['proteus/WaveTools.pyx'],
+<<<<<<< HEAD
```
Comment (smattis) [inline setup.py]: conflicts need resolved

**Comment by adimako** [comment]
This last commit reset all files apart from the WaveTools to the master version. Fingers crossed :)

**Comment by cekees** [comment]
For multiline docstrings you want to make the first line the "short" documentation (what it  does)  followed  by a blank line and then the complete documentations. Here's some good information: https://www.python.org/dev/peps/pep-0257/#id17

**Comment by cekees** [comment]
We're not pedantic on naming conventions but we try to use capital names for classes and lower case for  functions and objects. Here is a reasonable guide: https://www.python.org/dev/peps/pep-0008/#prescriptive-naming-conventions

File: proteus/WaveTools.pyx [hunk 8]
```diff
@@ -859,55 +944,55 @@ def w(self,x,z,t):
     plt.close("all")
     np.savetxt("Duck_series.txt",zip(rawtime,neweta))
 
+    for Nf in np.array([8,16,32,64,128]):
 
-    tseries = timeSeries(timeSeriesFile= "Duck_series.txt",
```
Comment (cekees) [inline proteus/WaveTools.pyx]: I believe timeSeries is a class name so it ought to be TimeSeries.

**Comment by adimako** [comment]
@cekees I replaced the DoublePeakedRandom waves with a new class because a): The two spectra used exactly the same phase arrays and this is not the case b) Well, I thought it would be nice to make it as general as possible. Now theoretically we can enter two completely different spectral functions, at any frequency range, different directions and reconstruct the time series. 

This may come in handy when modelling swell waves + wind waves or waves with 2nd order correction, or bidirectional sea states and in general where ever there is a clear separation of the spectral peaks in frequency and direction space

This may be also handy for the directional waves, as we can describe it as an addition of multiple spectra in different directions

**Comment by cekees** [comment]
Great! Thanks for doing that. I thought the DoublePeaked implementation was too much of a specialized hack as well. Glad you have made something more general.

File: proteus/WaveTools.pyx [hunk 9]
```diff
@@ -0,0 +1,1355 @@
...
+    :param depth: water depth [L]
+    :param g: gravity vector
+    :param vDir (vertical direction - opposite to the gravity vector)
+    :param comp: component "x", "y" or "z"
+    """
+
+    phase = x*kDir[0]+y*kDir[1]+z*kDir[2] - omega*t  + phi
+    Z =  (vDir[0]*x + vDir[1]*y+ vDir[2]*z) - mwl
+# We should not really be enforcing the velocity outside these bounds
+    if( (Z < -depth) or (Z > 2*amplitude)):
```
Comment (smattis) [inline proteus/WaveTools.pyx]: I think that this implementation makes it really hard (and impossible in some cases) to set boundary conditions. If you have a wave series with variable wave heights, the region in space that will raise this error changes for each amplitude. I can't just mark a section of the boundary and set a Dirichlet BC on that boundary.

**Comment by adimako** [comment]
@smattis Thanks for spotting this. I carried this from the regular wave implementation. Also the depth restriction only works if you give by accident a depth less than the actual depth of the domain. I will go ahead and remove it

**Comment by adimako** [comment]
@tridelat  I used rotate3D function for rotating the wave direction and it works!

**Comment by cekees** [comment]
I'm checking on why the tests are failing.

**Comment by cekees** [comment]
This PR depends on #292, so we should  probably wait to merge it. @tridelat it looks like #292 is almost ready. Is there anything I can do to help besides review  the PR?

**Comment by adimako** [comment]
test_wavetools fails because of matplotlib functions. I will remove these once I have debugged completely the direct reconstruction of the timeseries

**Comment by adimako** [comment]
@cekees I think we are good to merge now. There are few things yet to do

- 2nd order correction for nonlinear random waves
- Python notebook
- Have a think about how to distribute in space a single-point time series

**Comment by adimako** [comment]
@cekees On second thought, let me do one commit more because some of the class / funciton arguments are not well documented and then we can merge

**Comment by cekees** [comment]
@adimako We should try to get this ready to merge next week. It  depends on @tridelat's PR.

**Pull request merged**